### DISCUSSION TITLE
Questions about the route guard in the React-Router

### DISCUSSION THREAD

**Discussion by WDLieqi** [community/community#141247] [Programming Help] [answered]
I wonder if there's a way to implement permission verification for routing. For example, when I input `/admin/1`, the user does not have permission and needs to return NotFound, but the URL cannot be changed. It should still display `/admin/1`, rather than redirecting to `/notFound`.

Because the component to be verified is partially within child components, it is not feasible to change the rendering within the component itself, as it should display the entire page with the NotFound component.

I wonder if there's any way to achieve the desired effect.

**Comment by Mnahil-Naseer** [comment, marked as answer]
To implement permission verification in React while keeping the URL unchanged (e.g., /admin/1), you can use a Higher-Order Component (HOC) that conditionally renders the desired component or a "NotFound" component based on user permissions.

**Create a HOC for Permission Check:** This HOC checks user permissions and either renders the requested component or the NotFound component.

**Wrap Routes with HOC:** Use the HOC to wrap protected routes, ensuring that the URL stays the same even if the user doesn't have the required permissions.

**Comment by Berchez** [comment]
You can achieve this by using a combination of route guards and conditional rendering in your React application. Here’s a general approach using React Router:

1. Create a ProtectedRoute Component: This component will check if the user has the necessary permissions. If not, it will render the NotFound component while keeping the URL unchanged.
2. Use Context or State for Authentication: Store the user’s authentication status and permissions in a context or state.
3. Implement Conditional Rendering: Render the NotFound component if the user lacks the required permissions.

**Comment by WDLieqi** [comment]
Thanks for your answers. HOC was the answer I needed. In the end, I chose to recursively filter the Router configuration manually, which only required setting a hidden variable.
